libraries/Microsoft.Bot.Builder.AI.LUIS/LuisV3Util.cs

```diff
@@ -0,0 +1,193 @@
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.AI.Luis
+{
+    // Utility functions used to extract and transform data from Luis SDK
+    internal static class LuisV3Util
+    {
+        internal const string _metadataKey = "$instance";
+        internal const string _geoV2 = "builtin.geographyV2.";
+        internal static readonly HashSet<string> _dateSubtypes = new HashSet<string> { "date", "daterange", "datetime", "datetimerange", "duration", "set", "time", "timerange" };
+
+        internal static string NormalizedIntent(string intent) => intent.Replace('.', '_').Replace(' ', '_');
+
+        internal static IDictionary<string, IntentScore> GetIntents(JObject luisResult)
+        {
+            IDictionary<string, IntentScore> result = null;
+            var intents = (JObject)luisResult["intents"];
+            if (intents != null)
+            {
+                var dict = new Dictionary<string, IntentScore>();
+                foreach (var intent in intents)
+                {
+                    dict.Add(NormalizedIntent(intent.Key), new IntentScore { Score = intent.Value["score"] == null ? 0.0 : intent.Value["score"].Value<double>() });
+                }
+
+                result = dict;
+            }
+
+            return result;
+        }
+
+        internal static string NormalizedEntity(string entity)
+        {
+            // Type::Role -> Role
+            var type = entity.Split(':').Last();
+            return type.Replace('.', '_').Replace(' ', '_');
+        }
+
+        internal static void GeographyTypes(JToken source, Dictionary<string, string> geoTypes)
+        {
+            if (source != null)
+            {
+                if (source is JObject obj)
+                {
+                    if (obj.TryGetValue("type", out var type) && type.Type == JTokenType.String && type.Value<string>().StartsWith(_geoV2))
+                    {
+                        var path = type.Path.Replace(_metadataKey + ".", string.Empty);
+                        path = path.Substring(0, path.LastIndexOf('.'));
+                        geoTypes.Add(path, type.Value<string>().Substring(_geoV2.Length));
+                    }
+                    else
+                    {
+                        foreach (var property in obj.Properties())
+                        {
+                            GeographyTypes(property.Value, geoTypes);
+                        }
+                    }
+                }
+                else if (source is JArray arr)
+                {
+                    foreach (var elt in arr)
+                    {
+                        GeographyTypes(elt, geoTypes);
+                    }
+                }
+            }
+        }
+
+        internal static JToken MapProperties(JToken source, bool inInstance, Dictionary<string, string> geoTypes)
+        {
```

Comment:
This method is *really* hard to follow. Consider decomposing and turning into simpler methods. Correctness tests are probably needed, as it's quite complex.

Reply:
The oracle tests in effect test all of this since they compare the result of this, to the expected results.